### PR TITLE
Adding more distributions to contrib

### PR DESCRIPTION
**Pull request by neerajprad**
This adds the following distributions to contrib:
 - [x] binomial with probs, logits
 - [x] bernoulli with logits
 - [x] cauchy
 - [x] half cauchy
 - [x] exponential

As a follow-up to this PR, I'll add more tests for distribution args check validation.

### CODE REVIEW

File: numpyro/contrib/distributions/discrete.py [hunk 1]
```diff
@@ -20,23 +31,125 @@ def _to_logits_bernoulli(probs):
...
+            self._validate_sample(value)
+        dtype = get_dtypes(self.logits)[0]
+        value = lax.convert_element_type(value, dtype)
+        total_count = lax.convert_element_type(self.total_count, dtype)
+        log_factorial_n = gammaln(total_count + 1)
+        log_factorial_k = gammaln(value + 1)
+        log_factorial_nmk = gammaln(total_count - value + 1)
+        # Note that: torch.log1p(-self.probs)) = - torch.log1p(self.logits.exp()))
+        return (log_factorial_n - log_factorial_k - log_factorial_nmk +
+                value * self.logits - total_count * np.log1p(np.exp(self.logits)))
```
Comment (fehiepsi) [inline numpyro/contrib/distributions/discrete.py]: It might be better to address the overflow problem when logits is large. You can see [my implementation here](https://github.com/pyro-ppl/numpyro/pull/101/files#diff-fe7dcb1e0f5ffa51f99755b000e28ccbR87). But I am not sure if it resolves the discontinuity problem which we have in PyTorch previously.

Comment (neerajprad) [reply numpyro/contrib/distributions/discrete.py]: Makes sense, let me play around with your solution and put it in.

File: numpyro/distributions/transforms.py [hunk 2]
```diff
@@ -43,47 +44,19 @@ def log_abs_det_jacobian(self, x, y):
...
     def __call__(self, x):
-        for part in self.parts:
-            x = part(x)
-        return x
+        return np.abs(x)
 
-    def inv(self, y):
-        for part in self.parts[::-1]:
-            y = part.inv(y)
+    def _inv(self, y):
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/transforms.py]: Oh no..that was a mistake. I will revert this. Thanks for catching this.

File: numpyro/contrib/distributions/continuous.py [hunk 3]
```diff
@@ -2,56 +2,152 @@
...
+    arg_constraints = {'rate': constraints.positive}
+    support = constraints.positive
+
+    def __init__(self, rate, validate_args=None):
+        self.rate = rate
+        super(Exponential, self).__init__(batch_shape=np.shape(rate), validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        u = random.uniform(key, shape=size + self.batch_shape)
+        return np.log1p(-(-u)) / self.rate
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/contrib/distributions/continuous.py]: Meant to use `-log1p(-u)`. Will change, also added a test for mean/var that would detect this.
